### PR TITLE
Separate ignored backlog requests into separate list.

### PR DESCRIPTION
**Pull request by jberry-suse**
A useful distinction to make. Reason is displayed on hover.

Confirmed this works if:
- `dashboard` package does not exists
- `ignored_requests` file does not exits
- `ignored_requests` is empty
- `ignored_requests` is not empty

### CODE REVIEW

**Comment by jberry-suse** [comment]
@coolo Not sure if the PRs are visible so ping when you get a minute for this and others from me.

**Comment by coolo** [comment]
I see them

File: app/controllers/obs_factory/staging_projects_controller.rb [hunk 1]
```diff
@@ -16,6 +16,17 @@ def index
         format.html do
           @staging_projects = StagingProjectPresenter.sort(@distribution.staging_projects)
           @backlog_requests = Request.with_open_reviews_for(by_group: 'factory-staging', target_project: @distribution.name)
+          @ignored_requests = YAML.load("#{PackageFile.new(
```
Comment (coolo) [inline app/controllers/obs_factory/staging_projects_controller.rb]: this looks rather bizarre. why are putting things quoted in a string? Let me guess: you don't know about .to_s ?

File: app/views/obs_factory/staging_projects/_infos.html.haml [hunk 2]
```diff
@@ -26,3 +26,15 @@
         - if counter > 5 && @backlog_requests.size > counter + 1
           %li ... #{@backlog_requests.size - counter} more
           - break
+
+  - if !@backlog_requests_ignored.empty?
```
Comment (coolo) [inline app/views/obs_factory/staging_projects/_infos.html.haml]: !.empty? is present?

File: app/views/obs_factory/staging_projects/_infos.html.haml [hunk 3]
```diff
@@ -26,3 +26,15 @@
...
+    %p Ignored:
+    %ul{class: "staging_backlog_ignored"}
+      - counter = 0
```
Comment (coolo) [inline app/views/obs_factory/staging_projects/_infos.html.haml]: you can use each_with_index, no need to count yourself

File: app/views/obs_factory/staging_projects/_infos.html.haml (hunk 2 above)
Comment (bgeuken) [inline app/views/obs_factory/staging_projects/_infos.html.haml]: Not exactly. [present? is !blank?](http://api.rubyonrails.org/classes/Object.html#method-i-present-3F). But probably still what you want

File: app/controllers/obs_factory/staging_projects_controller.rb (hunk 1 above)
Comment (jberry-suse) [inline app/controllers/obs_factory/staging_projects_controller.rb]: All languages and communities are different. Many find calling toString() and similar tabo, but I'll change. I am not a ruby nor rails guy.

File: app/views/obs_factory/staging_projects/_infos.html.haml (hunk 3 above)
Comment (jberry-suse) [inline app/views/obs_factory/staging_projects/_infos.html.haml]: Just copied the code directly above it...

Comment (coolo) [inline app/views/obs_factory/staging_projects/_infos.html.haml]: Yeah, then happy refactoring :)

File: app/controllers/obs_factory/staging_projects_controller.rb (hunk 1 above)
Comment (jberry-suse) [inline app/controllers/obs_factory/staging_projects_controller.rb]: This issue here is how to handle `nil` in the case that no file is found? This is handle gracefully the way it is now, but a direct call to `to_s()` results in `no implicit conversion of nil into String` when the file is no present.

**Comment by jberry-suse** [comment]
Changes made and applied the refactoring `each_with_index` to existing code..

```diff
diff --git a/app/views/obs_factory/staging_projects/_infos.html.haml b/app/views/obs_factory/staging_projects/_infos.html.haml
index 581301c..d1a98b6 100644
--- a/app/views/obs_factory/staging_projects/_infos.html.haml
+++ b/app/views/obs_factory/staging_projects/_infos.html.haml
@@ -27,14 +27,12 @@
           %li ... #{@backlog_requests.size - counter} more
           - break
 
-  - if !@backlog_requests_ignored.empty?
+  - if @backlog_requests_ignored.present?
     %p Ignored:
     %ul{class: "staging_backlog_ignored"}
-      - counter = 0
-      - @backlog_requests_ignored.each do |req|
+      - @backlog_requests_ignored.each_with_index do |req,counter|
         %li
           = link_to elide(req.package, length = 19), main_app.request_show_path(req.number), title: @ignored_requests[req.number]
-        - counter += 1
         - if counter > 5 && @backlog_requests_ignored.size > counter + 1
           %li ... #{@backlog_requests_ignored.size - counter} more
           - break
```

File: app/controllers/obs_factory/staging_projects_controller.rb (hunk 1 above)
Comment (coolo) [inline app/controllers/obs_factory/staging_projects_controller.rb]: PackageFile.new returns nil? Wow - neat interface. Anyway, you better handle this case by setting @ignored_requests directly to some default value

Comment (jberry-suse) [inline app/controllers/obs_factory/staging_projects_controller.rb]:
```diff
diff --git a/app/controllers/obs_factory/staging_projects_controller.rb b/app/controllers/obs_factory/staging_projects_controller.rb
index a72de66..8faa24f 100644
--- a/app/controllers/obs_factory/staging_projects_controller.rb
+++ b/app/controllers/obs_factory/staging_projects_controller.rb
@@ -16,10 +16,13 @@ module ObsFactory
         format.html do
           @staging_projects = StagingProjectPresenter.sort(@distribution.staging_projects)
           @backlog_requests = Request.with_open_reviews_for(by_group: 'factory-staging', target_project: @distribution.name)
-          @ignored_requests = YAML.load("#{PackageFile.new(
+          file = PackageFile.new(
             project_name: "#{params[:project]}:Staging",
             package_name: "dashboard",
-            name: "ignored_requests")}")
+            name: "ignored_requests")
+          unless file.nil? or file.to_s.nil?
+            @ignored_requests = YAML.load(file.to_s)
+          end
           if !@ignored_requests.nil? and @ignored_requests
             @backlog_requests_ignored = @backlog_requests.select { |req| @ignored_requests.key?(req.number) }
             @backlog_requests = @backlog_requests.select { |req| !@ignored_requests.key?(req.number) }
```

Could also remove condition and just use `@ignored_requests = YAML.load("#{file}")`, but that is basically what existed before. I was incorrect in that new() returned `nil`, but rather that `to_s` returns `nil` which to me seems like a bug. Normally, a to string method always returns a string representation even if that is a blank string. Anyway this works although I am not sure it is really an improvement.

Comment (jberry-suse) [inline app/controllers/obs_factory/staging_projects_controller.rb]: Caused by this line, perhaps I should file a PR against OBS.
https://github.com/openSUSE/open-build-service/blob/3a554ed1c4e03862373358b0b9a733ea7f81aede/src/api/app/models/backend_file.rb#L60

Comment (jberry-suse) [inline app/controllers/obs_factory/staging_projects_controller.rb]: Confirmed, that changing the `nil` to `""` fixes the issue. I'll file a PR and see what is said. If changed the following works.

```ruby
            @ignored_requests = YAML.load(PackageFile.new(
            project_name: "#{params[:project]}:Staging",
            package_name: "dashboard",
            name: "ignored_requests").to_s)
```

Comment (jberry-suse) [inline app/controllers/obs_factory/staging_projects_controller.rb]: Created openSUSE/open-build-service#2872.

**Pull request merged**